pkg/syncer/controller.go

```diff
@@ -0,0 +1,443 @@
+		toClient:                  toClient,
+		fromClient:                fromClient,
+		direction:                 direction,
+		upstreamClusterName:       kcpClusterName,
+		mutators:                  make(mutatorGvrMap),
+		advancedSchedulingEnabled: advancedSchedulingEnabled,
+	}
+
+	if len(mutators) > 0 {
+		c.mutators = mutators
+	}
+
+	if direction == SyncDown {
+		c.upsertFn = c.applyToDownstream
+		c.deleteFn = c.deleteFromDownstream
+	} else {
+		c.upsertFn = c.updateStatusInUpstream
+		c.deleteFn = func(ctx context.Context, gvr schema.GroupVersionResource, namespace, name string) error {
+			if c.advancedSchedulingEnabled {
+				return ensureUpstreamFinalizerRemoved(ctx, gvr, c.toClient, namespace, c.workloadClusterName, c.upstreamClusterName, name)
+			}
+			return nil
+		}
+	}
+
+	for _, gvrstr := range gvrs {
+		gvr, _ := schema.ParseResourceArg(gvrstr)
+
+		fromInformers.ForResource(*gvr).Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.AddToQueue(*gvr, obj)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				oldUnstrob := oldObj.(*unstructured.Unstructured)
+				newUnstrob := newObj.(*unstructured.Unstructured)
+
+				if c.direction == SyncDown {
+					if !deepEqualApartFromStatus(oldUnstrob, newUnstrob) {
+						c.AddToQueue(*gvr, newUnstrob)
+					}
+				} else {
+					if !deepEqualFinalizersAndStatus(oldUnstrob, newUnstrob) {
+						c.AddToQueue(*gvr, newUnstrob)
+					}
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.AddToQueue(*gvr, obj)
+			},
+		})
+		klog.InfoS("Set up informer", "direction", c.direction, "clusterName", kcpClusterName, "pcluster", pcluster, "gvr", gvr)
+	}
+
+	c.fromInformers = fromInformers
+
+	return &c, nil
+}
+
+type holder struct {
+	gvr         schema.GroupVersionResource
+	clusterName logicalcluster.Name
+	namespace   string
+	name        string
+}
+
+func (c *Controller) AddToQueue(gvr schema.GroupVersionResource, obj interface{}) {
+	objToCheck := obj
+
+	tombstone, ok := objToCheck.(cache.DeletedFinalStateUnknown)
+	if ok {
+		objToCheck = tombstone.Obj
+	}
+
+	metaObj, err := meta.Accessor(objToCheck)
+	if err != nil {
+		klog.Errorf("%s: error getting meta for %T", c.name, obj)
+		return
+	}
+
+	qualifiedName := metaObj.GetName()
+	if len(metaObj.GetNamespace()) > 0 {
+		qualifiedName = metaObj.GetNamespace() + "/" + qualifiedName
+	}
+	if c.direction == SyncDown {
+		qualifiedName = metaObj.GetClusterName() + "|" + qualifiedName
+	}
+	klog.Infof("Syncer %s: adding %s %s to queue", c.name, gvr, qualifiedName)
+
+	c.queue.Add(
+		holder{
+			gvr:         gvr,
+			clusterName: logicalcluster.From(metaObj),
+			namespace:   metaObj.GetNamespace(),
+			name:        metaObj.GetName(),
+		},
+	)
+}
+
+// Start starts N worker processes processing work items.
+func (c *Controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.fromInformers.Start(ctx.Done())
```

Comment:
Am I right that the informer of the other side passed to either `NewSpecSyncer` or `NewStatusSyncer` is never started and waited for cache sync ?

Reply:
Good point. Updated.

Reply:
Do you still need to `Start` and `WaitForCacheSync` here now that it is done here: https://github.com/kcp-dev/kcp/pull/1011/files#diff-b043590efbb8609f4eced1377afbe4f51358c5c628fb79c0fdd24e8f9d553551R186 ?

Reply:
oh, lost that change during rebase it seems.

Reply:
fixed